### PR TITLE
check non_exhaustive attr and private fields for transparent types

### LINKED ISSUE #78586: repr(transparent) should not consider non_exhaustive types as 1-ZSTs outside their crate (`repr_transparent_external_private_fields`)
This tracks eventually removing support for `repr(transparent)` types ignoring ZST that might not actually remain ZST in the future or on other targets.

### Types that can gain more fields

In crate A:
```rust
#[non_exhaustive]
pub struct UnitType {}
```
In crate B:
```rust
#[repr(transparent)]
pub struct TransparentThing(i32, UnitType);
```

I expected that to fail in crate B, because the reason crate A put `#[non_exhaustive]` on the type was to allow it to get additional fields in the future, which would no longer allow it to work in `repr(transparent)`.

But this actually compiles today, creating what I would consider an accidental semver hazard.

https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Stability.20of.201-ZSTness/near/215120408

---

### repr(C) types

```rust
#[repr(C)]
struct CZst([u8; 0]);

#[repr(transparent)]
pub struct TransparentThing(i32, CZst);
```
This should fail since `CZst` actually is not a ZST for all C compilers, and since the type can be relevant for the ABI.

### PR DESCRIPTION
**Pull request by fee1-dead**
Fixes #78586.

### CODE REVIEW

**Comment by oli-obk** [comment]
I think we also need to address enums: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7cfa15a0f159c2aff5f070ac52f4d485

Also, I'm not actually sure what the expected behaviour of exporting `Baz` is:

```rust
#[non_exhaustive]
pub struct Foo {}

#[repr(transparent)]
pub struct Baz(pub Foo);
```

can `Baz` now be used as a ZST in another transparent type in another crate? That seems just as problematic as using the non-exhaustive type directly.

**Comment by fee1-dead** [comment]
> can `Baz` now be used as a ZST in another transparent type in another crate?

No, and this checks that as it recurses into the fields and error if any transitive field is non_exhaustive or has private fields.

**Comment by oli-obk** [comment]
> No, and this checks that as it recurses into the fields and error if any transitive field is non_exhaustive or has private fields.

cool, please add tests for that!

**Comment by fee1-dead** [comment]
There are several tests that already test this (See `ExternalIndirection` and `InternalIndirection`) although they are not repr(transparent)

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -3132,6 +3132,59 @@ declare_lint! {
     "detects unexpected names and values in `#[cfg]` conditions",
 }
 
+declare_lint! {
+    /// The `repr_transparent_external_private_fields` lint
+    /// detects types marked `#[repr(trasparent)]` that (transitively)
```
Comment (scottmcm) [inline compiler/rustc_lint_defs/src/builtin.rs:3137]: nit:
```suggestion
    /// detects types marked `#[repr(transparent)]` that (transitively)
```

**Comment by fee1-dead** [comment]
@bors r=oli-obk

minor doc change, no need to review again

**Pull request merged**